lib/css_parser/parser.rb

```diff
@@ -164,14 +164,44 @@ def add_block!(block, options = {})
       parse_block_into_rule_sets!(block, options)
     end
 
-    # Add a CSS rule by setting the +selectors+, +declarations+ and +media_types+.
+    # Add a CSS rule by setting the +selectors+, +declarations+
+    # and +media_types+. Optional pass +filename+ , +offset+ for source
+    # reference too.
     #
-    # +media_types+ can be a symbol or an array of symbols.
-    def add_rule!(selectors, declarations, media_types = :all)
-      rule_set = RuleSet.new(selectors, declarations)
-      add_rule_set!(rule_set, media_types)
-    rescue ArgumentError => e
-      raise e if @options[:rule_set_exceptions]
+    # +media_types+ can be a symbol or an array of symbols. default to :all
+    # optional fields for source location for source location
+    # +filename+ can be a string or uri pointing to the file or url location.
+    # +offset+ should be Range object representing the start and end byte locations where the rule was found in the file.
+    def add_rule!(*args, selectors: nil, block: nil, filename: nil, offset: nil, media_types: :all) # rubocop:disable Metrics/ParameterLists
+      if args.any?
```

Comment:
ideally not support args at all without `deprecated: true` being set, but still a good step forward

Reply:
Maybe I'm not understanding what your suggesting.
It would be a breaking change it we now require `deprecated: true` to pass keyword arguments. If somebody are doing the change to add the that keyword argument why don't they change all to keywords arguments.